ractor.rb

```diff
@@ -1,34 +1,180 @@
+#
+# In addition to that, an argument to Ractor.new would be passed to block and available there
+# as if received by Ractor.receive, and the last block value would be sent outside of the
+# ractor as if sent by Ractor.yield.
+#
+# A little demonstration on a classic ping-pong:
+#
+#     server = Ractor.new do
+#       puts "Server starts: #{self.inspect}"
+#       puts "Server sends: ping"
+#       Ractor.yield 'ping'                       # The server doesn't know the receiver and sends to whoever interested
+#       received = Ractor._receive                # The server doesn't know the sender and receives from whoever sent
+#       puts "Server received: #{received}"
+#     end
+#
+#     client = Ractor.new(server) do |srv|        # The server is sent inside client, and available as srv
+#       puts "Client starts: #{self.inspect}"
+#       received = srv.take                       # The Client takes a message specifically from the server
+#       puts "Client received from " \
+#            "#{srv.inspect}: #{received}"
+#       puts "Client sends to " \
+#            "#{srv.inspect}: pong"
+#       srv.send 'pong'                           # The client sends a message specifically to the server
+#     end
+#
+#     sleep(0.1)
+#
+# This will output:
+#
+#     Server starts: #<Ractor:#2 test.rb:1 running>
+#     Server sends: ping
+#     Client starts: #<Ractor:#3 test.rb:8 running>
+#     Client received from #<Ractor:#2 rac.rb:1 blocking>: ping
+#     Client sends to #<Ractor:#2 rac.rb:1 blocking>: pong
+#     Server received: pong
+#
+# It is said that Ractor receives messages via the <em>incoming port</em>, and sends them
+# to the <em>outgoing port</em>. Either one can be disabled with Ractor.close_incoming and
+# Ractor.close_outgoing respectively.
+#
+# == Shareable and unshareable objects
+#
+# When the object is sent to and from the ractor, it is important to understand whether the
+# object is shareable or not. Shareable objects are basically those which can be used by several
+# threads without compromising thread-safety; e.g. immutable ones. Ractor.shareable? allows to
+# check this, and Ractor.make_shareable tries to make object shareable if it is not.
+#
+#     Ractor.shareable?(1)            #=> true -- numbers and other immutable basic values are
+#     Ractor.shareable?('foo')        #=> false, unless the string is frozen due to # freeze_string_literals: true
+#     Ractor.shareable?('foo'.freeze) #=> true
+#
+#     str = 'foo'
+#     str.frozen?                 #=> false
+#     Ractor.make_shareable(str)
+#     str.frozen?                 #=> true
+#
+# When a shareable object is sent (via #send or Ractor.yield), no additional processing happens,
+# and it just becomes usable by both ractors. When an unshareable object is sent, it can be
+# either _copied_ or _moved_. The first is the default, and it makes the object's full copy by
+# deep cloning of its structure.
```

Comment:
Yeah, makes sense. I changed the whole explanation/example to:

When a shareable object is sent (via #send or Ractor.yield), no additional processing happens,
and it just becomes usable by both ractors. When an unshareable object is sent, it can be
either _copied_ or _moved_. The first is the default, and it makes the object's full copy by
deep cloning of non-shareable parts of its structure.

```ruby
data = ['foo', 'bar'.freeze]
r = Ractor.new{
 data2 = Ractor.receive
 puts "In ractor: #{data2.object_id}, #{data2[0].object_id}, #{data2[1].object_id}"
}
r.send(data)
sleep(0.1)
puts "Outside  : #{data.object_id}, #{data[0].object_id}, #{data[1].object_id}"
```

This will output:

```
In ractor: 340, 360, 320
Outside  : 380, 400, 320
```

(Note that object id of both array and non-frozen string inside array have changed inside
the ractor, showing it is different objects. But the second array's element, which is a
shareable frozen string, has the same object_id.)

Reply:
I like it